### PR TITLE
[tosa] Implement torch.linear support.

### PR DESCRIPTION
**Pull request by sjarus**
Refactor matmul into separate class and derive variants:
- matmul
- mm, bmm
- linear

Signed-off-by: Suraj Sudhir <suraj.sudhir@arm.com>

### CODE REVIEW

**Comment by sjarus** [comment]
All bit accuracy tests for matmul variants passing: 

Results PASS test_matmul_16_x16__f32
Results PASS test_matmul_16_x16x32__f32
Results PASS test_matmul_16x32_x32__f32
Results PASS test_matmul_16x32_x32x17__f32
Results PASS test_matmul_8x16x32_x32__f32
Results PASS test_matmul_8x16x32_x32x17__f32
Results PASS test_matmul_16x32_x8x32x17__f32
Results PASS test_matmul_16_x8x16x32__f32
Results PASS test_matmul_8x16x32_x8x32x17__f32
Results PASS test_matmul_4x8x16x32_x32__f32
Results PASS test_matmul_4x8x16x32_x32x17__f32
Results PASS test_matmul_4x8x16x32_x8x32x17__f32
Results PASS test_matmul_4x8x16x32_x1x32x17__f32
Results PASS test_matmul_16_x4x8x16x32__f32
Results PASS test_matmul_4x8x16x32_x4x8x32x17__f32
Results PASS test_matmul_4x8x16x32_x1x8x32x17__f32
Results PASS test_matmul_4x8x16x32_x1x1x32x17__f32
Results PASS test_mm_16x32_x32x17__f32
Results PASS test_bmm_8x16x32_x8x32x17__f32
Results PASS test_linear_5x3_f32
Results PASS test_linear_1x3_f32
PASS: 213, FAIL: 0

**Comment by sjarus** [comment]
@silvasean The build failure seems completely unrelated to anything in this push and I don't see it when I run 'cmake --build build --target check-torch-mlir' locally. Any idea what's happening ?

**Comment by silvasean** [comment]
> @silvasean The build failure seems completely unrelated to anything in this push and I don't see it when I run 'cmake --build build --target check-torch-mlir' locally. Any idea what's happening ?

It looks like an issue related to a newer version of PyTorch (and unrelated to your PR, though if you could fix it that would be much appreciated) -- the CI always uses the latest snapshot. You can try `python -m pip install --upgrade -r requirements.txt` to get the latest version of PT and reproduce the issue.

**Comment by cathyzhyi** [comment]
FYI, the build failure should be fixed by this PR https://github.com/llvm/torch-mlir/pull/536

**Comment by sjarus** [comment]
I just tried it; it does update deps but there's no build failure. 

Successfully installed certifi-2021.10.8 charset-normalizer-2.0.10 idna-3.3 pybind11-2.9.0 requests-2.27.1 torch-1.11.0.dev20211231+cpu torchvision-0.12.0.dev20211231+cpu urllib3-1.26.8 wheel-0.37.1

**Comment by sjarus** [comment]
> FYI, the build failure should be fixed by this PR #536

Thanks! Please feel free to look at this review too if you'd like - it just leverages the earlier matmul work to also implement torch.linear (just adding an rhs transpose and a bias add to the regular matmul process).

**Comment by sjarus** [comment]
I rebased after #536 was merged and it does fix this, thank you @cathyzhyi .

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -1358,20 +1352,218 @@ class ConvertAtenMatMulOp : public OpConversionPattern<AtenOpT> {
...
+    auto rhsTy = rhs.getType().cast<RankedTensorType>();
+
+    if (!lhsTy || !rhsTy)
+      return op.emitError("Only ranked tensor types supported in TOSA matmul");
+
+    auto lhsRank = lhsTy.getRank();
+    auto rhsRank = rhsTy.getRank();
+
+    if (lhsRank != 2 && lhsRank != 3)
+      return op.emitError("aten.Linear called but lhs rank not 2 or 3");
```
Comment (pashu123) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:1481]: NIT: Replace `lhs` and `rhs`  with `input` and `weight` respectively while emitting errors.

Comment (sjarus) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Done

**Comment by sjarus** [comment]
Nit feedback incorporated. Merging.

**Comment by cathyzhyi** [comment]
@sjarus Hey Suraj, sry I missed this in my review but some of the `matchAndRewrite` functions are not marked as `override` which result in compiler warning ` warning: 'matchAndRewrite' overrides a member function but is not marked 'override' [-Wsuggest-override]
  matchAndRewrite(AtenOpT op, OpAdaptor adaptor,` could you fix those?

**Comment by sjarus** [comment]
Sure, I'm working on Conv2D now, but will roll this in along with it . Would that work ?

**Comment by cathyzhyi** [comment]
> Sure, I'm working on Conv2D now, but will roll this in along with it . Would that work ?

yea, that would work! thanks!

**Pull request merged**